### PR TITLE
Fix ElectronContextMenuRenderer to use passed in anchor XY values

### PR DESCRIPTION
**Pull request by kenneth-marut-work**
Signed-off-by: Kenneth Marut <kenneth.marut@ericsson.com>
Signed-off-by: Colin Grant <colin.grant@ericsson.com>

#### What it does

Fixes: #7722

This PR addresses  #7722 which was uncovered in #7105 where the ElectronContextMenuRenderer did not utilize the passed in `anchor` parameter to render the menu at a given X/Y location. This caused the gear-icon and folder-scope context menus to be rendered at the mouse location rather than the passed-in anchor location. This issue is fixed by adding an identical `MouseEvent` check as in: https://github.com/eclipse-theia/theia/blob/cea62337848477b12b61ed7acf30634e12acc52d/packages/core/src/browser/menu/browser-context-menu-renderer.ts#L30

And also by rounding anchor XY values to integer forms (due to a conversion error down the line) and passing the arguments into `menu.popup({ x: Math.round(x), y: Math.round(y) });` in the `electron-menu-context-renderer.ts`

#### How to test

1. Pull latest changes from #7105, build, and run in Electron
2. Open new preferences UI with multiroot workspace
3. Click on gear-icon and/or folder scope tab in preferences UI, observe bad behavior of context menu
4. Pull changes from this PR, and rebuild in Electron
5. Relaunch and run steps 2-3
6. Observe context menu behaior

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts [hunk 1]
```diff
@@ -44,7 +44,9 @@ export class ElectronContextMenuRenderer extends ContextMenuRenderer {
 
     protected doRender({ menuPath, anchor, args, onHide }: RenderContextMenuOptions): ElectronContextMenuAccess {
         const menu = this.menuFactory.createContextMenu(menuPath, args);
-        menu.popup({});
+        const { x, y } = anchor instanceof MouseEvent ? { x: anchor.clientX, y: anchor.clientY } : anchor!;
+        // x and y values must be Ints or else there is a conversion error
+        menu.popup({ x: Math.round(x), y: Math.round(y) });
```
Comment (akosyakov) [inline packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts:49]: @kittaakos maybe you can help to test it a bit against Electron

Comment (kittaakos) [reply packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts:49]: I have tried it, clicked around the application: I did not notice any regression. Very nice contribution, thank you! 👍 (Note: I haven't had any issues with the context menu before either.)

@kenneth-marut-work, I do not get why you have to round the coordinates? `MouseEvent.clientX` has a [double floating-point value](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/clientX). Have you experienced any malfunction?

Even if the numbers are not integers, the `MouseEvent` is part of the [`contextmenu event`](https://developer.mozilla.org/en-US/docs/Web/API/Element/contextmenu_event), why do not we pass `x, y` further as is? I do not see any reason for the rounding.

Comment (kenneth-marut-work) [reply packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts:49]: Hey @kittaakos, thanks for the comment. I am not sure why this is necessary either, however if the number is not rounded I receive this error when testing (this has been the case on my colleague's machine as well):

image

Comment (kittaakos) [reply packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts:49]: I could reproduce it with this change:
```diff
diff --git a/packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts b/packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts
index b93ae2e67..e1a4f03b5 100644
--- a/packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts
+++ b/packages/core/src/electron-browser/menu/electron-context-menu-renderer.ts
@@ -46,7 +46,7 @@ export class ElectronContextMenuRenderer extends ContextMenuRenderer {
         const menu = this.menuFactory.createContextMenu(menuPath, args);
         const { x, y } = anchor instanceof MouseEvent ? { x: anchor.clientX, y: anchor.clientY } : anchor!;
         // x and y values must be Ints or else there is a conversion error
-        menu.popup({ x: Math.round(x), y: Math.round(y) });
+        menu.popup({ x: x + 0.125, y });
         // native context menu stops the event loop, so there is no keyboard events
         this.context.resetAltPressed();
         if (onHide) {
```

```
Uncaught Error: Could not call remote function 'anonymous'. Check that the function signature is correct. Underlying error: Error processing argument at index 1, conversion failure from 607.125

Error: Could not call remote function 'anonymous'. Check that the function signature is correct. Underlying error: Error processing argument at index 1, conversion failure from 607.125
    at callFunction (/Users/akos.kitta/gi…r/rpc-server.js:258)
    at /Users/akos.kitta/gi…r/rpc-server.js:409
    at EventEmitter.ipcMain.on.args (/Users/akos.kitta/gi…r/rpc-server.js:273)
    at EventEmitter.emit (events.js:182)
    at WebContents.<anonymous> (/Users/akos.kitta/gi…web-contents.js:368)
    at WebContents.emit (events.js:182)
    at callFunction (/Users/akos.kitta/gi…r/rpc-server.js:258)
    at /Users/akos.kitta/gi…r/rpc-server.js:409
    at EventEmitter.ipcMain.on.args (/Users/akos.kitta/gi…r/rpc-server.js:273)
    at EventEmitter.emit (events.js:182)
    at WebContents.<anonymous> (/Users/akos.kitta/gi…web-contents.js:368)
    at WebContents.emit (events.js:182)
    at errors.ts:26
```

The error comes from electron. I am OK with the rounding.

**Comment by kittaakos** [comment]
@akosyakov, do you have any remarks on this PR? Otherwise, I'm going to merge it.
The electron guys have completely rewritten the `rpc-server`, I think this `x, y` rounding becomes obsolete once we bump up the electron version:
 - `4.2.x`: https://github.com/electron/electron/blob/4-2-x/lib/browser/rpc-server.js
 - `master`: https://github.com/electron/electron/blob/master/lib/browser/rpc-server.js

**Pull request merged**